src/java/org/apache/cassandra/db/virtual/DescribeTables.java

```diff
@@ -0,0 +1,306 @@
+import org.apache.cassandra.db.marshal.CompositeType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.marshal.UserType;
+import org.apache.cassandra.dht.LocalPartitioner;
+import org.apache.cassandra.schema.Functions;
+import org.apache.cassandra.schema.IndexMetadata;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.schema.Types;
+import org.apache.cassandra.schema.ViewMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+public class DescribeTables
+{
+    private static final String KEYSPACE = "keyspace_name";
+    private static final String CQL = "cql";
+
+    private static final CompositeType utfComposite = CompositeType.getInstance(UTF8Type.instance, UTF8Type.instance);
+
+    public static Collection<VirtualTable> getAll(String name)
+    {
+        return ImmutableList.of(new DescribeKeyspaceTable(name),
+                                new DescribeIndexesTable(name),
+                                new DescribeTypesTable(name),
+                                new DescribeAggregatesTable(name),
+                                new DescribeFunctionsTable(name),
+                                new DescribeViewsTable(name),
+                                new DescribeTablesTable(name));
+    }
+
+    static final class DescribeKeyspaceTable extends AbstractVirtualTable
+    {
+        DescribeKeyspaceTable(String keyspace)
+        {
+            super(TableMetadata.builder(keyspace, "describe_keyspace")
+                               .comment("cql for keyspace metadata")
+                               .kind(TableMetadata.Kind.VIRTUAL)
+                               .partitioner(new LocalPartitioner(UTF8Type.instance))
+                               .addPartitionKeyColumn(KEYSPACE, UTF8Type.instance)
+                               .addRegularColumn(CQL, UTF8Type.instance)
+                               .build());
+        }
+
+        @Override
+        public DataSet data(DecoratedKey partitionKey)
+        {
+            String keyspace = UTF8Type.instance.compose(partitionKey.getKey());
+
+            SimpleDataSet result = new SimpleDataSet(metadata());
+            result.row(keyspace)
+                  .column(CQL, SchemaCQLHelper.getKeyspaceAsCQL(keyspace));
+            return result;
+        }
+
+        public DataSet data()
+        {
+            SimpleDataSet result = new SimpleDataSet(metadata());
+            for (String keyspace : Schema.instance.getKeyspaces())
+            {
+                result.row(keyspace)
+                      .column(CQL, SchemaCQLHelper.getKeyspaceAsCQL(keyspace));
+            }
+            return result;
+        }
+    }
+
+    static abstract class AbstractDescribeTable extends AbstractVirtualTable
+    {
+        AbstractDescribeTable(String keyspace, String name)
+        {
+            super(TableMetadata.builder(keyspace, "describe_" + name)
```

Comment:
Given all these tables will have the same schema, I think it might be more easily used if the “type” was just a column, rather than having to go to a whole new table. So you could just “select * from virtual_schema” and get everything.

Reply:
complexity there is there are things like having a UDT with same name as a table in the same keyspace. We can have something like ((type), keyspace, name) so that way we can provide uniqueness to everything and it would only be funny for keyspaces.